File: apps/api/languages.py
"""The one place the API decides what a language is.

The internal key is the ISO 639-1 code, matching `words.language` and every
RPC argument. English long names ('spanish') are a display concern and a
legacy input format; they are accepted at the boundary and converted here,
never passed further in.

This replaces six ad-hoc mapping tables that had drifted apart -- two of them
silently missing French, which is how a French learner ended up being served
Spanish words.
"""

from __future__ import annotations

# code -> English display name
LANGUAGES: dict[str, str] = {
    "es": "Spanish",
    "de": "German",
    "it": "Italian",
    "fr": "French",
}

SUPPORTED_CODES: tuple[str, ...] = tuple(LANGUAGES)

_BY_NAME = {name.lower(): code for code, name in LANGUAGES.items()}


def to_code(language: str | None) -> str | None:
    """Return the ISO code for a code or an English name, else None.

    None rather than a fallback: a bad value should produce an empty result,
    not silently the wrong language. The previous `default: 'spanish'` in the
    extension is exactly the bug this avoids.
    """
    if not language:
        return None
    normalized = language.strip().lower()
    if normalized in LANGUAGES:
        return normalized
    return _BY_NAME.get(normalized)


def require_code(language: str | None) -> str:
    """to_code(), raising on anything unrecognised."""
    code = to_code(language)
    if code is None:
        raise ValueError(
            f"Unsupported language: {language!r}. Expected one of {', '.join(SUPPORTED_CODES)}."
        )
    return code


def display_name(language: str | None) -> str | None:
    """Human-readable name, for UI and log lines only."""
    code = to_code(language)
    return LANGUAGES[code] if code else None

File: apps/web/src/lib/languages.ts
/**
 * The one place the web app decides what a language is.
 *
 * The internal key is the ISO 639-1 code, matching `words.language` and every
 * RPC argument. English names are for display only.
 *
 * Before this existed, eight hooks reached the database by sending
 * `language.name.toLowerCase()` — depending on the *display name* happening to
 * equal the database key. That coupling is why renaming a language in the UI
 * would have silently returned zero rows.
 */

export type LanguageCode = 'es' | 'de' | 'it' | 'fr';

export interface Language {
  code: LanguageCode;
  /** English display name. Never sent to the database. */
  name: string;
  /** flagcdn.com identifier — a *country* code, which is why it is stored
   *  separately rather than reusing `code`. */
  flag: string;
}

export const LANGUAGES: Record<LanguageCode, Language> = {
  es: { code: 'es', name: 'Spanish', flag: 'es' },
  de: { code: 'de', name: 'German', flag: 'de' },
  it: { code: 'it', name: 'Italian', flag: 'it' },
  fr: { code: 'fr', name: 'French', flag: 'fr' },
};

export const LANGUAGE_LIST: Language[] = Object.values(LANGUAGES);

const BY_NAME: Record<string, LanguageCode> = Object.fromEntries(
  LANGUAGE_LIST.map((l) => [l.name.toLowerCase(), l.code])
) as Record<string, LanguageCode>;

export function isLanguageCode(value: unknown): value is LanguageCode {
  return typeof value === 'string' && value in LANGUAGES;
}

/**
 * Accepts an ISO code or a legacy English name; returns null for anything else.
 *
 * Null rather than a fallback on purpose: an unrecognised value should produce
 * an empty result, not silently the wrong language.
 */
export function toLanguageCode(value: string | null | undefined): LanguageCode | null {
  if (!value) return null;
  const normalized = value.trim().toLowerCase();
  if (isLanguageCode(normalized)) return normalized;
  return BY_NAME[normalized] ?? null;
}

export function getLanguage(value: string | null | undefined): Language | null {
  const code = toLanguageCode(value);
  return code ? LANGUAGES[code] : null;
}
